Mathlib/Tactic/ComputeDegree.lean

```diff
@@ -0,0 +1,223 @@
+
+* Currently, the tactic does not handle goals of the form `degree f ≤ d`, where `d` is not the
+  coercion of a `Nat`.  This should be fixed.
+* Deal with equalities, instead of inequalities (i.e. implement `compute_degree`).
+* Add functionality to deal with (some) exponents that are not closed natural numbers.
+* Add support for proving goals of the from `natDegree f ≠ 0` and `degree f ≠ 0`.
+* Make sure that `degree` and `natDegree` are equally supported.
+
+##  Implementation details
+
+We start with a goal of the form `natDegree f ≤ d` or `degree f ≤ d`.
+Recurse into `f` breaking apart sums, products and powers.  Take care of numerals,
+`C a, X (^ n), monomial a n` separately.
+-/
+
+section Tactic
+
+open Polynomial
+
+namespace Mathlib.Tactic.ComputeDegree
+
+section mylemmas
+
+/-!
+###  Simple lemmas about `natDegree`
+
+The lemmas in this section deduce inequalities of the form `natDegree f ≤ d`, using
+inequalities of the same shape.
+This allows a recursive application of the `compute_degree_le` tactic on a goal,
+and on all the resulting subgoals.
+-/
+
+variable {R : Type _}
+
+section semiring
+variable [Semiring R]
+
+theorem add {a b : Nat} {f g : R[X]} (hf : natDegree f ≤ a) (hg : natDegree g ≤ b) :
+    natDegree (f + g) ≤ max a b :=
+(f.natDegree_add_le g).trans $ max_le_max ‹_› ‹_›
+
+theorem mul {a b : Nat} {f g : R[X]} (hf : natDegree f ≤ a) (hg : natDegree g ≤ b) :
+    natDegree (f * g) ≤ a + b :=
+(f.natDegree_mul_le).trans $ add_le_add ‹_› ‹_›
+
+theorem pow {a : Nat} (b : Nat) {f : R[X]} (hf : natDegree f ≤ a) :
+    natDegree (f ^ b) ≤ b * a :=
+natDegree_pow_le.trans (Nat.mul_le_mul rfl.le ‹_›)
+
+theorem C_le (a : R) : natDegree (C a) ≤ 0 := (natDegree_C a).le
+theorem nat_cast_le (n : Nat) : natDegree (n : R[X]) ≤ 0 := (natDegree_nat_cast _).le
+theorem zero_le : natDegree (0 : R[X]) ≤ 0 := natDegree_zero.le
+theorem one_le : natDegree (1 : R[X]) ≤ 0 := natDegree_one.le
+
+end semiring
+
+section ring
+variable [Ring R]
+
+theorem neg {a : Nat} {f : R[X]} (hf : natDegree f ≤ a) : natDegree (- f) ≤ a :=
+(natDegree_neg f).le.trans ‹_›
+
+theorem sub {a b : Nat} {f g : R[X]} (hf : natDegree f ≤ a) (hg : natDegree g ≤ b) :
+    natDegree (f - g) ≤ max a b :=
+(f.natDegree_sub_le g).trans $ max_le_max ‹_› ‹_›
+
+theorem int_cast_le (n : Int) : natDegree (n : R[X]) ≤ 0 := (natDegree_int_cast _).le
+
+end ring
+
+end mylemmas
+
+open Lean Meta Elab.Tactic
+
+/-!
+Four helper lemmas to build `Expr`essions: `mkMax, mkPow, mkNatDegree, mkDegree`.
+-/
+
+/-- Return `max a b` using `Max.max`. This method assumes `a` and `b` have the same type. -/
+def mkMax (a b : Expr) : MetaM Expr := mkAppM ``Max.max #[a, b]
+
+/-- Return `a ^ b` using `HPow.hPow`. -/
+def mkPow (a b : Expr) : MetaM Expr := mkAppM ``HPow.hPow #[a, b]
+
+/-- Returns `natDegree pol`. -/
+def mkNatDegree (pol : Expr) : MetaM Expr := mkAppM ``natDegree #[pol]
+
+/-- Returns `degree pol`. -/
+def mkDegree (pol : Expr) : MetaM Expr := mkAppM ``degree #[pol]
+
```

Comment:
```suggestion
```

The current version using `MVarId`s no longer needs these `Expr`-building lemmas.